apps/aeutil/lib/serialization.ex

```diff
@@ -8,28 +8,97 @@ defmodule Aeutil.Serialization do
-      prev_hash: bech32_binary(block.header.prev_hash, :header, direction),
-      txs_hash: bech32_binary(block.header.txs_hash, :txs, direction)}
-    new_txs = Enum.map(block.txs, fn(tx) -> tx(tx, direction) end)
-    Block.new(%{block | header: Header.new(new_header), txs: new_txs})
+    {new_header, new_txs} =
+      case direction do
+        :deserialize ->
+          built_header = Header.new(block["header"])
+          header = header(built_header, direction)
+          txs = Enum.map(block["txs"], fn(tx) ->
+              tx(tx, direction)
+            end)
+          {header, txs}
+        :serialize ->
+          header = header(block.header, direction)
+          txs = Enum.map(block.txs, fn(tx) ->
+              tx(tx, direction)
+            end)
+          {header, txs}
+      end
+    Block.new(header: new_header, txs: new_txs)
   end
 
-  @spec tx(SignedTx.t(), :serialize | :deserialize) :: SignedTx.t()
+  @spec header(Header.t(), :serialize | :deserialize) :: Header.t()
+  def header(header, direction) do
+    %{header |
+      chain_state_hash: bech32_binary(header.chain_state_hash, :chainstate, direction),
+      prev_hash: bech32_binary(header.prev_hash, :header, direction),
+      txs_hash: bech32_binary(header.txs_hash, :txs, direction)}
+  end
+
+  @spec tx(SignedTx.t() | map(), :serialize | :deserialize) :: SignedTx.t()
   def tx(tx, direction) do
-    new_data = %{tx.data |
-                 from_acc: hex_binary(tx.data.from_acc, direction),
-                 to_acc: hex_binary(tx.data.to_acc, direction)}
-    new_signature = base64_binary(tx.signature, direction)
-    %SignedTx{data: SpendTx.new(new_data), signature: new_signature}
+    {new_data, new_signature} =
+      case direction do
+        :deserialize ->
+          built_tx_data = build_tx_data(tx["data"])
+          data = serialize_tx_data(built_tx_data, direction)
+          signature = base64_binary(tx["signature"], direction)
+          {data, signature}
+        :serialize ->
+          data = serialize_tx_data(tx.data, direction)
+          signature = base64_binary(tx.signature, direction)
+          {data, signature}
+      end
+    %SignedTx{data: new_data, signature: new_signature}
+  end
+
+  @spec build_tx_data(map()) :: transaction_types()
+  def build_tx_data(tx_data) do
+    cond do
+      SignedTx.is_spend_tx(tx_data) ->
+        SpendTx.new(tx_data)
+      SignedTx.is_oracle_query_tx(tx_data) ->
+        OracleQueryTxData.new(tx_data)
+      SignedTx.is_oracle_registration_tx(tx_data) ->
+        OracleRegistrationTxData.new(tx_data)
+      SignedTx.is_oracle_response_tx(tx_data) ->
+        OracleResponseTxData.new(tx_data)
+    end
+  end
+
+  @spec serialize_tx_data(transaction_types(),
+                          :serialize | :deserialize) :: Serialization.transaction_types()
+  def serialize_tx_data(tx_data, direction) do
+    case tx_data do
+      %SpendTx{} ->
+        %{tx_data | from_acc: hex_binary(tx_data.from_acc, direction),
```

Comment:
I would suggest adding "type" field for easy deserialization. (Such field would be generally useful for anyone using our APIs)

Reply:
This doesn't seem like it would improve internal work

Reply:
@cytadela8 yes I thought of this as well, I will prepare a task to do this